apollo-runtime/src/appleMain/kotlin/com/apollographql/apollo3/network/http/StreamingNSURLSessionHttpEngine.kt

```diff
@@ -0,0 +1,313 @@
+import platform.posix.pthread_mutex_init
+import platform.posix.pthread_mutex_lock
+import platform.posix.pthread_mutex_t
+import platform.posix.pthread_mutex_unlock
+
+/**
+ * An [HttpEngine] based on [NSURLSession] with the ability to stream data as it is received when using `Transfer-Encoding: Chunked`.
+ * This is useful when using `@defer`.
+ *
+ * Note: this implementation only works when using the [new native memory manager](https://github.com/JetBrains/kotlin/blob/master/kotlin-native/NEW_MM.md)
+ * as multiple threads mutate the internal state.
+ *
+ * That is why it is considered experimental and subject to change.
+ *
+ * Using the default [DefaultHttpEngine] will work when using `@defer`, but the payloads are buffered until the response is fully
+ * received, thus losing the incremental nature of the response.
+ */
+@ApolloExperimental
+class StreamingNSURLSessionHttpEngine(
+    private val timeoutMillis: Long = 60_000,
+) : HttpEngine {
+
+  private val delegate = StreamingDataDelegate()
+
+  private val nsUrlSession = NSURLSession.sessionWithConfiguration(
+      configuration = NSURLSessionConfiguration.defaultSessionConfiguration(),
+      delegate = delegate,
+      delegateQueue = null
+  )
+
+  @Suppress("UNCHECKED_CAST")
+  override suspend fun execute(request: HttpRequest) = suspendAndResumeOnMain<HttpResponse> { mainContinuation, invokeOnCancellation ->
+    assertMainThreadOnNative()
+
+    val nsMutableURLRequest = NSMutableURLRequest.requestWithURL(
+        URL = NSURL(string = request.url)
+    ).apply {
+      setTimeoutInterval(timeoutMillis.toDouble() / 1000)
+
+      request.headers.forEach {
+        setValue(it.value, forHTTPHeaderField = it.name)
+      }
+
+      if (request.method == HttpMethod.Get) {
+        setHTTPMethod("GET")
+      } else {
+        setHTTPMethod("POST")
+        val requestBody = request.body
+        if (requestBody != null) {
+          setValue(requestBody.contentType, forHTTPHeaderField = "Content-Type")
+
+          if (requestBody.contentLength >= 0) {
+            setValue(requestBody.contentLength.toString(), forHTTPHeaderField = "Content-Length")
+          }
+          val body = Buffer().apply { requestBody.writeTo(this) }.readByteArray().toNSData()
+          setHTTPBody(body)
+        }
+      }
+      setCachePolicy(NSURLRequestReloadIgnoringCacheData)
+    }
+
+    val httpDataPipe = Pipe()
+    val httpDataSink = httpDataPipe.sink.buffer()
+    val httpDataSource = httpDataPipe.source.buffer()
+
+    val handler = object : StreamingDataDelegate.Handler {
+      override fun onResponse(response: NSHTTPURLResponse) {
+        mainContinuation.resumeWith(
+            buildHttpResponse(
+                data = httpDataSource,
+                httpResponse = response,
+                error = null,
+            )
+        )
+      }
+
+      override fun onData(data: NSData) {
+        httpDataSink.write(data.toByteString())
+        httpDataSink.flush()
```

Comment:
Out of curiosity, is this called on every chunk? Every x kB? Something else? How much do we need to flush?

Reply:
From what I see in my tests, it's called once per chunk, so flushing here makes the payload available to the reader

Reply:
Makes sense 👍 . Maybe add a comment up there?